crypto/rsa/rsa_pk1.c

```diff
@@ -248,8 +248,6 @@ int RSA_padding_check_PKCS1_type_2(unsigned char *to, int tlen,
     }
 
     OPENSSL_clear_free(em, num);
-    RSAerr(RSA_F_RSA_PADDING_CHECK_PKCS1_TYPE_2, RSA_R_PKCS_DECODING_ERROR);
-    err_clear_last_constant_time(1 & good);
```

Comment:
Note this and the below are API breaks. OpenSSL usually adds to the error queue on error conditions. This produces a strange one where it reports an error condition but doesn't add to the error queue.

It is especially bizarre because other error paths through RSA decrypt (internal malloc failure) *do* push to the error queue (it is perfectly fine for them to do so as those are uncorrelated with the input string).

The root problem here is, as I noted in https://github.com/openssl/openssl/pull/8446#issuecomment-471660131, confusion about success/failure conditions. `RSA_private_decrypt(RSA_PKCS1_PADDING)` has a failure condition. The problem is not that you all need to fix the implementation of it. The problem is that it is a failure condition in the first place. That API is simply incompatible with avoiding a padding oracle. You need a new one, or for callers to do the unpad themselves, as libssl currently does.

Indeed I'll draw your addition to the block just above this, lines 238 through 248. Even with this PR, this function is not constant-time with respect to that padding oracle. It does not have a consistent memory access pattern.

Reply:
There is no guarantee that all errors push something on the stack, for instance when there was
no memory to allocate the error state.  Applications that depend on that are broken.
But I am concerned about creating a padding oracle when the error queue contains a
RSA_R_PKCS_DECODING_ERROR from here, and later a different function fails, and pushes
some more errors on the error queue, then this padding error might be printed together with other
errors by using ERR_print_errors, nobody expects such a sensitive information from the error queue.

Reply:
Indeed, but there is no saving this function because it fundamentally has the wrong failure condition. Hackily papering over the usual caller expectations this won't change that fact.

The end result will be that this function subtly violates calling conventions, doesn't actually avoid a padding oracle (memory access pattern leaks), and yet it is still not safely callable in the usual way, because the caller needs to ignore the failure anyway, which is not a common API pattern.

It is better to cut through all this, add a different safer API, and discourage use of this one.

Reply:
I could fix the memory access pattern, but it would take O(N^2) accesses, would that be
worth it?

Reply:
O(N log N) should be possible via tricks like these:
https://boringssl.googlesource.com/boringssl/+/4ca8d131d316b6f072f546c98c1011bbecb3f8b2/crypto/cipher_extra/tls_cbc.c#160
https://boringssl.googlesource.com/boringssl/+/4ca8d131d316b6f072f546c98c1011bbecb3f8b2/crypto/fipsmodule/bn/shift.c#172

I don't think this is worth it, no, but I have been the one advocating giving up on this API. :-P

That this needs extra machinery further demonstrates why. In the only instances where the "replace with random" workaround for PKCS#1 v1.5 is viable, the caller already knows the expected length, so the constant-time checks can be much tighter. A better API would pass that length in. And without that better API, the caller needs to compare the return against the expected length, a further difficulty in applying this API to the workaround.

Reply:
So far the proposed new API is designed to work only with TLS.
I have a different application in mind, where I would like to check a non-constant
payload length in constant time, that is OPC/UA UserNamePassword encryption,
where the payload is variable length, with  4-byte length + password + nonce.
It would be sufficient to check that the 4-byte length is equal to mlen-4,
but that of course assumes that we are able to get the padding
removed in constant time what it is not as it seems, although the
the patch in 1.1.1b already broke applications (mine for instance :-( ).

Reply:
What exactly does this application do with the ciphertext end-to-end? In particular, what exactly are the inputs and failure conditions?

The TLS workaround for PKCS#1 v1.5 decryption is not immediately applicable to any arbitrary protocol. Again, the story is not that PKCS#1 v1.5 needs to be constant-time. It's that that decryption algorithm is fatally broken. It happens that, for TLS's use, we have a workaround which effectively folds RSA decryption and subsequent record layer MAC / Finished check into one operation. It is not a general fix. (E.g. the workaround failed SSLv2 because, IIRC, there was a server message in the way and the attacker could detect the non-determinism after brute-forcing a low-entropy export secret.)

Reply:
The plain text must be well formed, that means it has the length of the password and
the server chosen nonce in the beginning, if a chosen cipher text is decrypted
the chances are 2^32:1 that this check of the length field will fail.
Then it has at the end the nonce, 32 bytes, which was designed to protect against
replay, but it works also perfectly against chosen cipher text attacks, since the
likelihood to get the 32 bytes right, is as low as the OAEP to accept a random packet.
If any of those checks fail, we return the same error code as if the password was
wrong.  At least the initial 4 bytes need to be checked in constant time, ideally
also the nonce.

Reply:
I see. The length's 2^32 is usually on the small side for cryptography. I assume the 32-bit nonce is chosen by the decrypter and not the encrypter? That's more reasonable, but I don't think you can sanely check it with this API because it shifts it by the length of the output. Rather, it seems you want to decrypt with no padding, so that the nonce is in a public position where you can check it cleanly.

At the end of the day, remember that PKCS#1 v1.5 decryption is broken. This sort of thing is just workarounds. You should make sure your system has more solid options available, e.g. modern TLS has switched to using the key for signing, so it can eventually be retired.